client/js/widgets/widget.js

```diff
@@ -453,6 +453,37 @@ export class Widget extends StateManaged {
         }
       }
 
+      if(a.func == 'CANVAS') {
+        setDefaults(a, { mode: 'reset', x: 0, y: 0 });
+
+        if([ 'color', 'reset', 'setPixel' ].indexOf(a.mode) == -1)
+          problems.push(`Warning: Mode ${a.mode} will be interpreted as reset.`);
+
+        const execute = async function(widget) {
+          if(widget.get('type') == 'canvas') {
+            if(a.mode == 'setPixel')
+              await widget.setPixel(a.x, a.y, a.colorIndex);
+            else if(a.mode == 'color')
+              await widget.set('activeColor', a.colorIndex === undefined ? null : a.colorIndex);
+            else
+              await widget.reset();
+          }
+        };
```

Comment:
Maybe that function should be divided into two.  The color side of the function doesn't really have much to do with reset and set and the input is different
There could be more features added related to the color like previous or next color and change to the color map. (mode: set/inc/dec/change, value : number, color:"color string")
Also, I think that there should be an import and export function that puts the whole image into one compressed string. it is kinda impossible to set a good image at the moment.

Reply:
Here is my suggestions for CANVAS modes:
- cycleColor: signed integer indicating how many colors to move (default mode, value defaults to 1)
- setColor: set activeColor index (optionally set the mapped color)
- addColor: append a new color to colorMap
- clear: set all regions to "0"
- setPixel: set the color of a pixel
- getPixel: set a variable to the color index of a pixel
- getRegion: set a variable to the decompressed representation of a region
- setRegion: apply a decompressed representation of a region